### ISSUE TITLE
Any suggestions for handling 3rd party keyboards?

### ISSUE THREAD

**Issue by davidtlee** [driftyco/ionic-plugin-keyboard#115]
I realized that the callbacks only work for native keyboard. With iOS 8's 3rd party keyboard support, are there any suggested workarounds to make sure apps still work for those phones? (right now, the keyboard is covering the UI since I'm not able to resize the container since the callback doesn't get triggered)

For example, is there a way to detect when someone is using a 3rd party keyboard?

**Comment by beckson** [comment]
i also had these problem, could not find any solution.

**Comment by Hirbod** [comment]
I had the same problems. I've disabled third party keyboards for my app. I know, this is not a very good way, but it was the only way for me to keep the bugs away.

**Comment by davidtlee** [comment]
Oh neat! I didn't know it was possible to disable third party keyboards -- how did you do that?

**Comment by Hirbod** [comment]
Just put this inside of your AppDelegate

```obj-c
- (BOOL)application:(UIApplication *)application shouldAllowExtensionPointIdentifier:(NSString *)extensionPointIdentifier {
    if ([extensionPointIdentifier isEqualToString: UIApplicationKeyboardExtensionPointIdentifier]) {
        return NO;
    }
    return YES;
}
```

**Comment by beckson** [comment]
yeah, i had tried these code. But disabled third party keyboards is not a very good way, since too much people had used it and that must be changed their habits on mobile.

**Comment by Hirbod** [comment]
But actually, I've tested third party keyboard on iOS 8.4 and everyhing was working perfectly right now? :S. I guess the code is (at least for me) not necessary anymore

**Comment by Hirbod** [comment]
removed my last message, as it was non-sense. Yes @beckson , you're right, but there is now good way currently. Keyboard is the most problem on cordova for me..

**Comment by davidtlee** [comment]
Got it, thanks guys! :)

On Tue, Jul 21, 2015 at 9:22 PM, Hirbod <notifications@github.com> wrote:

> removed my last message, as it was non-sense. Yes @beckson
> <https://github.com/beckson> , you're right, but there is now good way
> currently. Keyboard is the most problem on cordova for me..
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/driftyco/ionic-plugin-keyboard/issues/115#issuecomment-123559420>
> .
>



-- 
David Lee
@davidtlee

**Comment by silasgarrison** [comment]
All you need to do is add an observer for when the keyboard frame changes.  Add:
```objectivec
 _keyboardChangeObserver
```
to line ~5 of IonicKeyboard.h then add:
```objectivec
_keyboardChangeObserver = [nc addObserverForName:UIKeyboardDidChangeFrameNotification
                               object:nil
                               queue:[NSOperationQueue mainQueue]
                               usingBlock:^(NSNotification* notification) {
                                   CGRect keyboardFrame = [notification.userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue];
                                   keyboardFrame = [self.viewController.view convertRect:keyboardFrame fromView:nil];
                                   
                                   [weakSelf.commandDelegate evalJs:[NSString stringWithFormat:@"cordova.fireWindowEvent('native.keyboardchange', { 'keyboardHeight': %@ }); ", [@(keyboardFrame.size.height) stringValue]]];

                                   //deprecated
                                   [weakSelf.commandDelegate evalJs:[NSString stringWithFormat:@"cordova.fireWindowEvent('native.changekeyboard', { 'keyboardHeight': %@ }); ", [@(keyboardFrame.size.height) stringValue]]];
                               }];
```
To line ~35 of IonicKeyboard.m.  Then all you have to do is add a listener to your JS:
```javascript
window.addEventListener(
	"native.keyboardchange",
	function(evt){
		// Oh sweet, now my keyboard height is > evt.keyboardHeight
	}
);
```

I use this to detect changes when someone hide's the typeahead suggestions or uses an Emoji keyboard.  It should work with all other 3rd party keyboards as well.

**Comment by davidtlee** [comment]
This looks really great! Thanks! I'll put it on my todo list of things to add support for. Glad there's a solution :)

**Comment by mikewagz** [comment]
Hi, does this plugin now support SwiftKey and other 3rd-party keyboards?
